ISISComputingGroup/ibex_gui
# Ticket4910 2 bundle python with gui

### Description of work
Added to the build process of ibex and the script generator a robocopy to bundle the latest build of python with the products of those build processes.

Edited the PreferenceSupplier to provide the developers python if not built or the bundled python if built.

Reversed the dependencies of ibex.ui, ibex.log and ibex.alarm in order to get the script generator to build without pulling in lots of unnecessary dependencies.

Created pipelines to test the new builds on the build server.

### Ticket

https://github.com/ISISComputingGroup/IBEX/issues/4910

### Acceptance criteria

- [x] Python is bundled with the ibex gui and the script generator standalone on build
- [x] The ibex gui and sript gen standalone use the bundled python
- [x] Developers use the developer version of python
- [x] The builds pass on the build server (ibex_gui_python_bundle_pipeline, script_gen_python_bundle_pipeline)

### Unit tests

It is very difficult to unit test as the process relies on using a specific project's resources. No unit tests provided.

### To test (Local build):

- Set up
  - I have added logging to the PreferenceSupplier.java to log which python is being used, we must remove this at the end of review
- [x] Run script gen standalone as dev
  - Check out branch and run scriptgenerator.product from eclipse
  - Confirm the developers python is being used
  - Confirm script gen functionality is working #4165 #4166
- [x] Run script gen in IBEX as dev
  - Remove hidden perspective from base/uk.ac.stfc.isis.ibex.e4.client/plugin_customization.ini in first line change from `uk.ac.stfc.isis.ibex.preferences/perspectives_not_shown=uk.ac.stfc.isis.ibex.client.e4.product.perspective.scriptGenerator` to `uk.ac.stfc.isis.ibex.preferences/perspectives_not_shown=`
  - Run ibex.product from eclipse
  - Confirm you can use the functionality of the script generator as expected from past tickets #4165 #4166
  - Confirm the developers python is being used
  - Add hidden perspective back
- [ ] Build script gen
  - Run build_script_generator.bat
  - Confirm python is copied from build into the target
  - Confirm you can use the functionality of the script generator as expected from past tickets #4165 #4166
  - Confirm it uses the bundled version of python (look for logs in workspace/logs/isis.log)
- [ ] Build ibex
  - Remove hidden perspective from base/uk.ac.stfc.isis.ibex.e4.client/plugin_customization.ini in first line change from `uk.ac.stfc.isis.ibex.preferences/perspectives_not_shown=uk.ac.stfc.isis.ibex.client.e4.product.perspective.scriptGenerator` to `uk.ac.stfc.isis.ibex.preferences/perspectives_not_shown=`
  - Run build.bat
  - Confirm python is copied from build into the target
  - Confirm you can use the functionality of the script generator as expected from past tickets #4165 #4166
  - Confirm it uses the bundled version of python (look for logs in workspace/logs/isis.log)
  - Add hidden perspective back
- Tear down
  - Remove logging from PreferenceSupplier.java

### To test (build servers):

- Confirm the two pipelines are building correctly (inspect console output and pull built versions to run on machine)
  - steps
    - [ ] Inspect console output to see build and copy have worked
    - [ ] Pull built version onto machine and run
    - [ ] Confirm they're using the correct python
  - pipelines
    - [ ] ibex_gui_python_bundle_pipeline
    - [ ] script_gen_python_bundle_pipeline

### System tests

No system tests added.

### Documentation

*Highlight and provide a link to any additions or changes to the documentation, if applicable. The aim is provide information to help the reviewer*

---

#### Code Review


### Final Steps
